fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/GraphExpansion.java

```diff
@@ -0,0 +1,298 @@
+ * a builder in an ordinary builder pattern and this class also provides getters to all elements and the ability to
+ * merge other objects of this class in a additive manner.
+ *
+ * This class also hides some data cleansing process, in particular related to potential duplicity on place holders,
+ * which must be corrected (distincted) before the place holders are used in the predicates of e.g. a match candidate.
+ */
+public class GraphExpansion implements KeyExpressionVisitor.Result {
+    @Nonnull
+    private final List<Value> resultValues;
+
+    @Nonnull
+    private final List<QueryPredicate> predicates;
+
+    @Nonnull
+    private final List<Quantifier> quantifiers;
+
+    @Nonnull
+    private final List<Placeholder> placeholders;
+
+    private GraphExpansion(@Nonnull final List<? extends Value> resultValues,
+                           @Nonnull final List<? extends QueryPredicate> predicates,
+                           @Nonnull final List<? extends Quantifier> quantifiers,
+                           @Nonnull final List<? extends Placeholder> placeholders) {
+        this.resultValues = ImmutableList.copyOf(resultValues);
+        this.predicates = ImmutableList.copyOf(predicates);
+        this.quantifiers = ImmutableList.copyOf(quantifiers);
+        this.placeholders = ImmutableList.copyOf(placeholders);
+    }
+
+    @Nonnull
+    public List<Value> getResultValues() {
+        return resultValues;
+    }
+
+    @Nonnull
+    public List<QueryPredicate> getPredicates() {
+        return predicates;
+    }
+
+    @Nonnull
+    public List<Quantifier> getQuantifiers() {
+        return quantifiers;
+    }
+
+    @Nonnull
+    public List<Placeholder> getPlaceholders() {
+        return placeholders;
+    }
+
+    @Nonnull
+    public List<CorrelationIdentifier> getPlaceholderAliases() {
+        return placeholders
+                .stream()
+                .map(ValueComparisonRangePredicate.Placeholder::getParameterAlias)
+                .collect(ImmutableList.toImmutableList());
+    }
+
+    @Nonnull
+    public QueryPredicate asAndPredicate() {
+        return AndPredicate.and(getPredicates());
+    }
+
+    @Nonnull
+    public GraphExpansion withPredicate(@Nonnull final QueryPredicate predicate) {
+        return new GraphExpansion(this.resultValues, ImmutableList.of(predicate), this.quantifiers, this.placeholders);
+    }
+
+    /**
+     * Method to <em>seal</em> a graph expansion in an instance of {@link Sealed}. A sealed graph expansion is immutable
+     * and can only be used to (repeatedly) build actual expressions.
+     * A graph expansion may contain duplicate information that have been added to it by callers. That is allowed and
+     * supported. In fact, in most cases, duplicates among e.g. {@link QueryPredicate}s come from merging individual
+     * {@link GraphExpansion}s into more complex ones. This method normalizes all elements before in order to return a
+     * sealed version of itself.
```

Comment:
This doesn't seem to be precisely true. In the interesting cases, it makes local placeholders and predicates.